context based solution

danielberndt opened the pull request:
This should be everything that is required to make a context based solution work.

notes:
 - the readme still needs to be updated
 - due to the changes of how owner vs parent based context is treated, I had to update the peerDep of react from 0.13 to 0.14
 - `<RadioGroup>` will render a `div` containing its children by default, but the wrapper component can be modified via passing e.g. a `comp="ul"` prop   
    this allows `RadioGroup` to have more than a single child node (see the example.jsx)

All in all I find this solution much easier on the end user's side. Plus it fixes #12!

chenglou commented on index.jsx:
```diff
@@ -33,12 +38,29 @@ export default React.createClass({
     onChange: PropTypes.func,
-    children: PropTypes.func.isRequired,
+    children: PropTypes.node.isRequired,
+    comp: PropTypes.oneOfType([
```
Humm not too sure about this. Is this necessary?

danielberndt replied:
What are you referring to? The existence of the `comp` prop, or the PropType definition? I assume you mean the former:

I've seen it quite often that libs offer a `comp` prop to allow customising the tag for the container. A common use case would be to allow using custom layout containers like `<Box>` or `<Flex>` or <InlineBlock>` or however people wanna call them.

An alternative approach would be that `<RadioGroup>` does not render a tag itself but only it's child. I feel that this is somewhat more inconvenient to use since users might expect that `<RadioGroup>` should be able to render multiple children.

chenglou replied:
I guess this is fine. Can you change the name to `Component` though?

nkbt replied:
I would also make it `defaultProp` rather then default value in render method.

chenglou commented:
Yeah seems legit!

I personally don't use this module anymore, but since you seem to, I've added you to the repo contributors and npm pushers list. Feel free to merge this and publish (and update the README).

Btw, shouldComponentUpdate would simply take context into consideration, right?

danielberndt commented:
Cool thanks! :)

Regarding your question on `shouldComponentUpdate` I'm not sure I get what situation you are pointing to. However here's a use case that's gonna be problematic, I think:

```
<RadioGroup selectedValue={foo}>
  <Box>
    <Radio value="apple"/> Apple
    <Radio value="orange"/> Orange
  </Box>
</RadioGroup>
```

if `<Box>` implements `shouldComponentUpdate` the default way, i.e. `shallowEqual(this.props, nextProps) && shallowEqual(this.state, nextState)`, changing the `selectedValue` will not propagate through to the `<Radio>`s because no props have changed for `<Box>`.

There's this [big issue](https://github.com/facebook/react/issues/2517) talking about this, but it's still an ongoing conversation.

I suggest that if we should go down the context-based path, to put a warning into the readme that this library might break if you use any components between `<RadioGroup>` and `<Radio>` that implement `shouldComponentUpdate`. (which seems unlikely, yet it's probably gonna happen)

nkbt commented:
Looks good to me (with comments about naming and defaultProps).

danielberndt commented:
Changed the `comp` prop to `Component` and updated the README.md. I did not change the links within the README pointing to the code yet, since we don't have a permalink for the commits yet.

Please let me know what you think!

chenglou commented:
Great. I'll go ahead and merge this then! Thanks a lot.
If you want to publish this, feel free to do so. The scripts are still there (although not as modern. They still work I think? @nkbt knows all about this =]). Feel free to close relevant issues too.

danielberndt commented:
I just packaged everything up and released v3.0.0, it seems however that the build is broken 😞 

The umd build is still assuming one default export whereas now there are two named exports. I'll research to see how to update the configuration.

danielberndt commented:
Okay, I more or less copied react-router's publishing method and released a working version v3.0.1. Not sure whether to unpublish v3.0.0 because that version won't help anyone...

nkbt commented:
Don't unpublish please, not good anyway...
Is it ok now?

nkbt commented:
Tip: to avoid problems with publishing always check with `npm install ~/path/to/lib` somewhere else and check if all is good.

danielberndt commented:
Thanks for the tip! Bookmarked it :)

installing the newest version worked fine on my application, but there seemed to be a problem here: #31 let's see what their answer is.

axelson commented:
I just wanted to thank you for this new API, much nicer to use! 👍
